File: typescript/src/tools/projects/getProjects.ts
import { ProjectGetAllSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = ProjectGetAllSchema;

type Params = z.infer<typeof schema>;

export const getProjectsHandler = async (context: Context, _params: Params) => {
	const orgId = await context.getOrgID();
	const projectsResult = await context.api.organizations().projects({ orgId }).list();

	if (!projectsResult.success) {
		throw new Error(`Failed to get projects: ${projectsResult.error.message}`);
	}

	return {
		content: [{ type: "text", text: JSON.stringify(projectsResult.data) }],
	};
};

const definition = getToolDefinition("projects-get");

const tool = (): Tool<typeof schema> => ({
	name: "projects-get",
	description: definition.description,
	schema,
	handler: getProjectsHandler,
});

export default tool;

File: python/tools/projects/get_projects.py
import json

from api.client import is_error, is_success
from schema.tool_inputs import ProjectGetAllSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def get_projects_handler(context: Context, _params: ProjectGetAllSchema) -> ToolResult:
    org_id = await context.get_org_id()
    projects_result = await context.api.organizations().projects(org_id).list()

    if is_error(projects_result):
        raise Exception(f"Failed to get projects: {projects_result.error}")

    assert is_success(projects_result)

    projects_data = [project.model_dump() for project in projects_result.data]

    return ToolResult(content=json.dumps(projects_data))


def get_projects_tool() -> Tool[ProjectGetAllSchema]:
    definition = get_tool_definition("projects-get")

    return Tool(
        name="projects-get",
        description=definition["description"],
        schema=ProjectGetAllSchema,
        handler=get_projects_handler,
    )
